diff -rupN a/utils/haddock/haddock-api/src/Haddock/Interface/Create.hs b/utils/haddock/haddock-api/src/Haddock/Interface/Create.hs
--- a/utils/haddock/haddock-api/src/Haddock/Interface/Create.hs	2026-01-27 22:05:51.043300487 +0000
+++ b/utils/haddock/haddock-api/src/Haddock/Interface/Create.hs	2026-01-27 22:06:07.467905246 +0000
@@ -176,7 +176,15 @@ createInterface1 flags unit_state mod_su
   -- See Note [Exporting built-in items]
   let builtinTys = DsiSectionHeading 1 (WithHsDocIdentifiers (mkGeneratedHsDocString "Builtin syntax") [])
       bonus_ds mods
-        | mdl == gHC_PRIM   = [ builtinTys, DsiExports funAvail ] <> mods
+       | mdl == gHC_PRIM =
+            [ builtinTys
+            , DsiExports $
+                {- Haddock does not want to sort avails, the order should be
+                    deterministically /derived from the source/.
+                   In this particular case, sorting funAvail would be a no-op anyway. -}
+                DefinitelyDeterministicAvails
+                  funAvail
+            ] <> mods
         | otherwise         = mods
 
   let
@@ -411,11 +419,11 @@ mkExportItems
           Just hsDoc' -> do
             doc <- processDocStringParas dflags pkgName hsDoc'
             pure [ExportDoc doc]
-      DsiExports avails ->
+      DsiExports (DetOrdAvails avails) ->
         -- TODO: We probably don't need nubAvails here.
         -- mkDocStructureFromExportList already uses it.
         concat <$> traverse availExport (nubAvails avails)
-      DsiModExport mod_names avails -> do
+      DsiModExport mod_names (DetOrdAvails avails) -> do
         -- only consider exporting a module if we are sure we are really
         -- exporting the whole module and not some subset.
         (unrestricted_mods, remaining_avails) <- unrestrictedModExports dflags thisMod modMap instIfaceMap avails (NE.toList mod_names)
